pkg/cmd/roachprod/install/cockroach.go

```diff
@@ -396,19 +396,86 @@ func (h *crdbInstallHelper) generateStartCmd(
 	binary := cockroachNodeBinary(h.c, nodes[nodeIdx])
 	keyCmd := h.generateKeyCmd(nodeIdx, extraArgs)
 
-	// NB: this is awkward as when the process fails, the test runner will show an
-	// unhelpful empty error (since everything has been redirected away). This is
-	// unfortunately equally awkward to address.
+	if h.c.IsLocal() {
+		// We cannot rely on sudo or systemd on the local machine; run using the
+		// background flag.
+		cmd := fmt.Sprintf(`
+			ulimit -c unlimited; mkdir -p %[1]s;
+			echo ">>> roachprod start: $(date)" >> %[1]s/roachprod.log;
+			ps axeww -o pid -o command >> %[1]s/roachprod.log;
+			[ -x /usr/bin/lslocks ] && /usr/bin/lslocks >> %[1]s/roachprod.log; %[2]s
+			export ROACHPROD=%[3]d%[4]s;
+			GOTRACEBACK=crash COCKROACH_SKIP_ENABLING_DIAGNOSTIC_REPORTING=1 %[5]s \
+			%[6]s %[7]s --background %[8]s >> %[1]s/cockroach.stdout.log \
+											         2>> %[1]s/cockroach.stderr.log \
+				|| (x=$?; cat %[1]s/cockroach.stderr.log; exit $x)`,
+			logDir,                  // [1]
+			keyCmd,                  // [2]
+			nodes[nodeIdx],          // [3]
+			h.c.Tag,                 // [4]
+			h.getEnvVars(),          // [5]
+			binary,                  // [6]
+			startCmd,                // [7]
+			strings.Join(args, " "), // [8]
+		)
+		return cmd, nil
+	}
+
+	// We run cockroach as a systemd service unit. This allows us to apply a
+	// strict memory limit and provides tools to monitor the process.
 	cmd := fmt.Sprintf(`
-		ulimit -c unlimited; mkdir -p %[1]s;
-		echo ">>> roachprod start: $(date)" >> %[1]s/roachprod.log;
-		ps axeww -o pid -o command >> %[1]s/roachprod.log;
-		[ -x /usr/bin/lslocks ] && /usr/bin/lslocks >> %[1]s/roachprod.log; %[2]s
-		export ROACHPROD=%[3]d%[4]s;
-		GOTRACEBACK=crash COCKROACH_SKIP_ENABLING_DIAGNOSTIC_REPORTING=1 %[5]s \
-		%[6]s %[7]s %[8]s >> %[1]s/cockroach.stdout.log \
-		                 2>> %[1]s/cockroach.stderr.log \
-			|| (x=$?; cat %[1]s/cockroach.stderr.log; exit $x)`,
+sudo systemctl reset-failed cockroach
+if systemctl is-active -q cockroach; then
+  echo "cockroach service already active"
+	echo "To get more information: systemctl status cockroach"
+	exit 1
+fi
+
+# The first time we run, install a small script that shows some helpful
+# information when we ssh in.
+if [ ! -e ${HOME}/.profile-cockroach ]; then
+  cat > ${HOME}/.profile-cockroach <<'EOF'
+echo ""
+if systemctl is-active -q cockroach; then
+	echo "cockroach is running; see: systemctl status cockroach"
+elif systemctl is-failed -q cockroach; then
+	echo "cockroach failed; see: systemctl status cockroach"
+else
+	echo "cockroach stopped"
+fi
+echo ""
+EOF
+  echo ". ${HOME}/.profile-cockroach" >> ${HOME}/.profile
+fi
+
+uid=$(id -u)
+gid=$(id -g)
+cat > ${HOME}/run-cockroach.sh <<'EOF' && \
+sudo systemd-run --unit cockroach --uid $uid --gid $gid -p MemoryMax=%[9]s bash ${HOME}/run-cockroach.sh
```

Comment:
Some more useful flags that we may want to add
-r, --remain-after-exit
--working-directory=

I also wonder, don't we need some probe to determine when to return? `--background` returns when the ports are bound (so that you don't get spurious connection refused when connecting right after `c.Start()`). What prevents that flake here?

Reply:
```
--user

    Talk to the service manager of the calling user, rather than the service manager of the system.
```

Maybe this avoids the need for going through `root`?

Reply:
I tried that, the memory limit doesn't work with --user.

Reply:
Good question about the probe.. Are you sure it's not the same currently? AFAICT with `--background` we fork at the very beginning: https://github.com/cockroachdb/cockroach/blob/master/pkg/cli/start.go#L283